Eliminates duplication of code in the controllers; fixes#74

alexander-kh opened the pull request:
- Common methods related to uploading of codes were reworked and moved from the vendors_controller.rb into Import module.
- Corresponding methods from Vendor model were moved into Offeror module.
- import method have been updated in order to reduce complexity.
- upload_page.html.erb templates for both offerors have been refactored.
- RSpec and Cucumber tests have been updated in accordance with changes.

Uzzsen commented:
Hi Alex,

The idea of taking out common reusable code into a separate module is great! 
I only wonder if there is a particular reason for the upload_page actions not taken into the Import module:

def upload_page _# in ProvidersController_
    @redeemifyCodes = current_provider.redeemifyCodes.all
end

 def upload_page _# in VendorsController_
    @vendorcodes = current_vendor.vendorCodes.all
 end

Also since the module methods deal with manipulation (not only Import) of codes and history updates why not naming it something like Codes or In_Out_Codes ?

alexander-kh commented:
Hi Oleg,

Thank you for reviewing!
I've moved it in Import module.
Both instance variables are not used in the template, but your remark was the starting point for the revision of other actions.
I've updated [template](https://github.com/strawberrycanyon/redeemify/pull/75/files#diff-c2337c89348946fab5f88ee0f6cbcb9f) associated with view_codes action in vendors_controller.rb, but there is a lot of work with other templates.

Yes, you are right, the module deals with the management of codes, and also with history.
I think 'Import' conveys the essence pretty well. but we can change it at any time.

tansaku commented on app/controllers/application_controller.rb:
```diff
@@ -18,6 +18,22 @@ def current_vendor
   end
   
+  def current_offeror
+    if params[:controller] == "providers"
```
could shorten with ternary? e.g.

```
params[:controller] == 'providers' ? current_provider : current_vender
```

maybe?

alexander-kh replied:
Yes, thanks! I'm also for conciseness.

tansaku commented on app/controllers/application_controller.rb:
```diff
@@ -18,6 +18,22 @@ def current_vendor
+  end
+  
+  def offeror_codes
+    if params[:controller] == "providers"
```
use ternary as above?

alexander-kh replied:
Done.

tansaku commented on app/controllers/concerns/import.rb:
```diff
@@ -0,0 +1,64 @@
+
+  def home
+    @histories_array = []
+    if current_offeror.history != nil
```
could `Offeror.home_set` be rewritten so that it returns a `[]` if it receives a nil?  or we could wrap `current_offeror.history` in `Array()`?

Lot's of ideas from Avdi Grimm's confident ruby relevant here ...

alexander-kh replied:
Thanks for pointing out. I rewrote Offeror.home_set
```
def self.home_set(histories)
  if histories.blank?
    return []
  else
    ...
    return histories_array
  end
end
```
import.rb:
```
def home
  @histories_array = Offeror.home_set(current_offeror.history)
  ...
end
```

tansaku commented on app/models/concerns/offeror.rb:
```diff
@@ -0,0 +1,99 @@
+module Offeror
+
```
let's ensure consistent 2 space indentation :-)

alexander-kh replied:
Perhaps GitHub represents it this way because originally indents were made by space.

tansaku commented on app/models/concerns/offeror.rb:
```diff
@@ -0,0 +1,99 @@
+module Offeror
+
+	def self.import(file, current, comment)
```
some very long methods - could we split out sub-methods and give them self-documenting names?

alexander-kh replied:
Agree. I've extracted two methods from the original one, but it can be refactored more thoroughly.
Can we create a separate GitHub issue for this task?

tansaku commented on app/models/concerns/offeror.rb:
```diff
@@ -0,0 +1,99 @@
+	private
+	
+  def self.add_code(offeror, code)
+    if offeror.class == Vendor
```
not so rubyish to check for class - wonder if we can find a duck typing alternative?

alexander-kh replied:
Agree.
```
if offeror.is_a? Vendor
```

tansaku commented on app/models/vendor.rb:
```diff
@@ -11,113 +11,20 @@ def defaultValue
 	end
 
+	def self.update_profile_vendor(current_vendor, info)
+		current_vendor.update_attributes(:cashValue => info["cashValue"],
```
indentations issues again ...

alexander-kh replied:
Thanks. I've updated every line manually, as in the case above.

tansaku commented:
@alexander-kh great work here - I made some inline comments - see what you think :-)

tansaku commented on app/controllers/application_controller.rb:
```diff
@@ -18,6 +18,15 @@ def current_vendor
   end
   
+  def current_offeror
+    params[:controller] == "providers" ? current_provider : current_vendor
```
so now we could DRY out across these two methods extracting

```rb
def providers?
  params[:controller] == 'providers'
end
```

but only if you feel like it :-)

tansaku commented on app/models/concerns/offeror.rb:
```diff
@@ -0,0 +1,102 @@
+    serialize_errors[:err_file] = file_check(file.path)
+    return serialize_errors if serialize_errors[:err_file]
+    f = File.open(file.path, "r")
+      f.each_line do |row|
```
maybe this anonymous block is a good place to extract a method?

tansaku commented on app/models/concerns/offeror.rb:
```diff
@@ -0,0 +1,102 @@
+  end
+  
+  def self.remove_unclaimed_codes(current)
+    if current.is_a? Vendor
```
could pull out three internal methods here to self document?

tansaku commented:
great work @alexander-kh a few more comments added - let me know if this is too much ... :-)

alexander-kh commented:
@tansaku, take a look, please:
[app/controllers/application_controller.rb](https://github.com/strawberrycanyon/redeemify/pull/75/files#diff-55c5b7aecfb519d0e4880eaf2788eb6e)
[app/models/concerns/offeror.rb](https://github.com/strawberrycanyon/redeemify/pull/75/files#diff-3493d3d410b6bb89f5ab228c8f955fc5)

tansaku commented:
great work @alexander-kh 

I think there's still some room to split out shorter methods, but I understand if you've reached a limit on working on this.  Maybe the thing to do is to open some refactoring tickets, and/or set up codeclimate monitoring on the repo?

Do you know when/if the team can make one of the scrums?

alexander-kh commented:
@tansaku, we can continue refactoring.

alexander-kh commented:
@tansaku,
Look at new [offeror.rb](https://github.com/strawberrycanyon/redeemify/pull/75/files#diff-3493d3d410b6bb89f5ab228c8f955fc5), please.
Class methods are instance methods now and invoking by the current_offeror in the [controllers](https://github.com/strawberrycanyon/redeemify/pull/75/files#diff-6cdca9aa432d2a6fa5a70306577e7baa).
RSpec tests were rewritten in accordance with changes:
- [providers_controller_spec.rb](https://github.com/strawberrycanyon/redeemify/pull/75/files#diff-bde9d2a55037a063d55702a97a691a8a)
- [vendors_controller_spec.rb](https://github.com/strawberrycanyon/redeemify/pull/75/files#diff-e7993e27c4aed880e65083eaf57aa093)
- [vendor_spec.rb](https://github.com/strawberrycanyon/redeemify/pull/75/files#diff-6fe6324dd46d95d04745c1e66216ee25)
